src/v/datalake/translation/scheduling.cc

```diff
@@ -0,0 +1,369 @@
+    auto holder = gate.hold();
+    vassert(!state._wait_hook.is_linked(), "Translator in waiting state");
+    vassert(state._running_hook.is_linked(), "Translator not in running state");
+    vassert(!state._force_flush_in_progress, "Duplicate force flush request");
+    vlog(datalake_log.debug, "Force flushing translator: {}", state);
+    state._force_flush_in_progress = true;
+    ssx::spawn_with_gate(gate, [&state, backoff] {
+        return state._translator->force_flush(backoff)
+          .then([&state]() { state._running_hook.unlink(); })
+          .finally([&state] { state._force_flush_in_progress = false; });
+    });
+}
+
+scheduler::scheduler(
+  size_t total_memory,
+  size_t memory_block_size,
+  std::unique_ptr<scheduling_policy> policy)
+  : _scheduling_policy(std::move(policy))
+  , _mem_tracker(reservations_tracker::make_default(
+      total_memory, memory_block_size, *this)) {
+    ssx::repeat_until_gate_closed_or_aborted(_state.gate, _state.as, [this] {
+        return main().handle_exception([](const std::exception_ptr& e) {
+            if (!ssx::is_shutdown_exception(e)) {
+                vlog(
+                  datalake_log.warn,
+                  "Encountered exception in main loop: {}, ignoring",
+                  e);
+            }
+        });
+    });
+}
+
+void scheduler::notify_ready(const translator_id& id) {
+    auto holder = _state.gate.hold();
+    vlog(datalake_log.trace, "ready notification from translator: {}", id);
+    auto it = _state.translators.find(id);
+    if (it == _state.translators.end() || it->second._wait_hook.is_linked()) {
+        return;
+    }
+    vlog(datalake_log.trace, "Marking the translator ready: {}", it->second);
+    it->second._running_hook.unlink();
+    _state.waiting.push_back(it->second);
+    _state_changed_cvar.signal();
+}
+
+bool scheduler::requires_scheduling_actions() const {
+    return !_state.waiting.empty() || _mem_tracker->memory_exhausted();
+}
+
+void scheduler::notify_memory_exhausted() {
+    vlog(datalake_log.debug, "memory exhausted notification");
+    _state_changed_cvar.signal();
+}
+
+ss::future<> scheduler::stop() {
+    vlog(datalake_log.debug, "Stopping scheduler");
+    _state.as.request_abort();
+    _state_changed_cvar.broken();
+    co_await _state.gate.close();
+    co_await ss::max_concurrent_for_each(
+      _state.translators, 32, [](auto& it) mutable {
+          return it.second._translator->stop();
+      });
+}
+
+ss::future<> scheduler::add_translator(std::unique_ptr<translator> translator) {
+    auto holder = _state.gate.hold();
+    const auto& id = translator->id();
+    const auto& translators = _state.translators;
+    vlog(
+      datalake_log.trace,
+      "request to add translator with id: {}, current_translators: {}",
+      id,
+      translators.size());
+    vassert(
+      !translators.contains(id),
+      "Duplicate translator registration: {}",
+      translator->id());
```

Comment:
Just a precaution, maybe we should return a bool instead of crashing? I guess this is relying on the datalake manager to only ever add one at a time, but that's a pretty distant invariant to rely on